### PR TITLE
bugfix removing multiple fieldsets does't work

### PR DESCRIPTION
**Pull request by svycka**
at the same time iterating collection items and removing causes problem:
When we have 3 items but we delete second and third. It fails and does not remove third element. So we have to move remove out of foreach

### CODE REVIEW

**Comment by svycka** [comment]
fixes #7443

**Comment by Maks3w** [comment]
Please add a unit test

**Comment by Maks3w** [comment]
So the issue is modify and read the iterator (PriorityList) in the same loop.

Should this to be fixed in PriorityList instead?

**Comment by svycka** [comment]
@Maks3w I don't think this should be allowed in PriorityList. Also don't see how this could be implemented.
```
foreach PriorityList
    remove
    remove
endforeach
PriorityList->commit
```
does not feel right

**Comment by svycka** [comment]
@Maks3w if you don't like how it is fixed I could change:
```php
foreach ($this->iterator as $name => $elementOrFieldset) {
// to
foreach ($this->iterator->getIterator() as $name => $elementOrFieldset) {
```
this would also fix the issue.

**Comment by Maks3w** [comment]
I delegate to other @zendframework/community-review-team member for their opinions.

To have a temportal `$removed` array is weird for me

**Comment by svycka** [comment]
I have found commit witch introduced the bug 
PR: #6240 
commit: https://github.com/zendframework/zf2/commit/9a5dd6055120e50e8103fc8d037679edee2b48c1#diff-31fe6580a729ea0519189ac9ef45928aL212

File: library/Zend/Form/Element/Collection.php [hunk 1]
```diff
@@ -221,6 +222,10 @@ public function populateValues($data)
                 ));
             }
 
+            $removed[] = $name;
+        }
+
+        foreach ($removed as $name) {
```
Comment (weierophinney) [inline library/Zend/Form/Element/Collection.php]: Why not just remove as we discover valid names?

Comment (weierophinney) [inline library/Zend/Form/Element/Collection.php]: Interestingly, because doing so causes unexpected behavior.

Comment (Maks3w) [inline library/Zend/Form/Element/Collection.php]: IMO this should be fixed in PriorityList instead

Comment (weierophinney) [inline library/Zend/Form/Element/Collection.php]: I've tried a few different approaches at this time. The problem is that doing removal while inside the loop leads to inconsistent results; the underlying iterator is not updated in place correctly. As such, the first pass is to determine which we can remove, and then we remove each.

**Pull request merged**